### PR TITLE
Apply limit to per-room display names when setting them

### PR DESCRIPTION
**Pull request by Azrenbeth**
Apply limit to per-room display names when setting them

Fixes issue #9236

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -49,6 +49,7 @@
 from synapse.util.distributor import user_left_room
 
 from ._base import BaseHandler
+from .profile import MAX_DISPLAYNAME_LEN
```
Comment (clokep) [inline synapse/handlers/room_member.py:52]: I think we generally prefer absolute imports now.

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -79,7 +80,7 @@ def __init__(self, hs: "HomeServer"):
         self.account_data_handler = hs.get_account_data_handler()
         self.event_auth_handler = hs.get_event_auth_handler()
 
-        self.member_linearizer = Linearizer(name="member")
+        self.member_linearizer: Linearizer = Linearizer(name="member")
```
Comment (clokep) [inline synapse/handlers/room_member.py:83]: This type hint shouldn't be necessary since it is declared as a `Linearizer`. 😕

File: synapse/handlers/room_member.py [hunk 3]
```diff
@@ -548,13 +549,17 @@ async def update_membership_locked(
         )
 
         if (
-            not self.allow_per_room_profiles and not is_requester_server_notices_user
-        ) or requester.shadow_banned:
+            (not self.allow_per_room_profiles and not is_requester_server_notices_user)
+            or requester.shadow_banned
+            or len(content.get("displayname", "")) > MAX_DISPLAYNAME_LEN
```
Comment (clokep) [inline synapse/handlers/room_member.py:554]: We should probably check the avatar URL while we're here.

I also don't think this check is correct -- if the display name is too long, this shouldn't mean that we also strip off the avatar URL! (And vice versa.)

@anoadragon453 asserted in #9236 that we should "fail the request". This is silently stripping it, which I'm unsure is the right thing to do.

File: synapse/handlers/room_member.py [hunk 4]
```diff
@@ -548,13 +549,17 @@ async def update_membership_locked(
...
+        ):
             # Strip profile data, knowing that new profile data will be added to the
             # event's content in event_creation_handler.create_event() using the target's
             # global profile.
             content.pop("displayname", None)
             content.pop("avatar_url", None)
+            # content may now be emtpy
+            content_specified = len(content) == 0
```
Comment (clokep) [inline synapse/handlers/room_member.py:562]: Why is this change necessary?

**Review by clokep — CHANGES_REQUESTED** [review]
I think this is on the right track, but I'm not sure about a couple of the changes.

File: synapse/handlers/room_member.py (hunk 2 above)
Comment (Azrenbeth) [reply synapse/handlers/room_member.py:83]: Not having that gave mypy errors about not being able to determine the type of `member_linearizer`. I wasn't able to track down what was making mypy unhappy so thought adding in the type hint was a good enough solution

File: synapse/handlers/room_member.py (hunk 3 above)
Comment (Azrenbeth) [reply synapse/handlers/room_member.py]: Have added the avatar URL length check

Was unsure if it was safe to "fail the request" - would that essentially mean having synapse prevent the event from ever having happened? What would happen if other servers did accept it?

File: synapse/handlers/room_member.py (hunk 4 above)
Comment (Azrenbeth) [reply synapse/handlers/room_member.py]: It wasn't necessary, I'd just added it as I thought it was weird that content_specified might be incorrect. I've deleted it though as it wasn't really relevant to this

File: changelog.d/10654.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Apply display name length limit to per-room display names.
```
Comment (clokep) [inline changelog.d/10654.bugfix:1]:
```suggestion
Enforce the maximum length for per-room display names and avatar URLs.
```

File: synapse/handlers/room_member.py [hunk 6]
```diff
@@ -556,6 +557,12 @@ async def update_membership_locked(
             content.pop("displayname", None)
             content.pop("avatar_url", None)
 
+        if len(content.get("displayname") or "") > MAX_DISPLAYNAME_LEN:
+            raise SynapseError(403, "Displayname is too long (max 256)")
```
Comment (clokep) [inline synapse/handlers/room_member.py:561]: For each of these, please use a `400` error with a code of `Codes.BAD_JSON`.

File: synapse/handlers/room_member.py (hunk 2 above)
Comment (clokep) [reply synapse/handlers/room_member.py:83]: I *think* there's something weird going on here because it is an abstract class. Seems like a reasonable enough solution.

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
@Azrenbeth Looks good! Any idea if it would easy to add tests for this?

**Comment by Azrenbeth** [comment]
> @Azrenbeth Looks good! Any idea if it would easy to add tests for this?

I don't really know how I could do that sorry!

**Comment by clokep** [comment]
> > @Azrenbeth Looks good! Any idea if it would easy to add tests for this?
> 
> I don't really know how I could do that sorry!

Taking a look I don't see a straightforward way to do this unfortunately. 😢
